types/src/validator_verifier.rs

```diff
@@ -159,54 +172,115 @@ impl ValidatorVerifier {
+            .expect("Failed to get multi signature")
+            .verify(message, &aggregated_key)
+            .map_err(|_| VerifyError::FailedToVerifyMultiSignature)?;
+        Ok(aggregated_sig)
+    }
+
+    /// This function will successfully return when at least quorum_size signatures of known authors
+    /// are successfully verified. It creates an aggregated public key using the voter bitmask passed
+    /// in the multi-signature and verifies the message passed in the multi-signature using the aggregated
+    /// public key.
+    ///
+    pub fn verify_multi_signatures<T: CryptoHash + Serialize>(
+        &self,
+        message: &T,
+        multi_signature: &MultiSignature,
     ) -> std::result::Result<(), VerifyError> {
-        self.check_num_of_signatures(aggregated_signature)?;
-        self.check_voting_power(aggregated_signature.keys())?;
-        let keys_and_signatures: Vec<(bls12381::PublicKey, bls12381::Signature)> =
-            aggregated_signature
-                .iter()
-                .flat_map(|(address, signature)| {
-                    let sig = signature.clone();
-                    self.get_public_key(address).map(|pub_key| (pub_key, sig))
-                })
-                .collect();
-        // Fallback is required to identify the source of the problem if batching fails.
-        if bls12381::Signature::batch_verify(message, keys_and_signatures).is_err() {
-            self.verify_aggregated_struct_signature(message, aggregated_signature)?
+        // Verify the number of signature is not greater than expected.
+        self.check_num_of_signatures(multi_signature)?;
+        let authors: HashSet<AccountAddress> = multi_signature
+            .get_voter_addresses(&self.get_ordered_account_addresses_iter().collect_vec())
+            .into_iter()
+            .collect();
+        // Verify the quorum voting power of the authors
+        self.check_voting_power(authors.iter())?;
+        #[cfg(any(test, feature = "fuzzing"))]
+        {
+            if self.quorum_voting_power == 0 {
+                // This should happen only in case of tests.
+                // TODO(skedia): Clean up the test behaviors to not rely on empty signature
+                // verification
+                return Ok(());
+            }
         }
+        // Verify the optimistically aggregated signature.
+        let pub_keys_to_agg: Vec<&PublicKey> = self
+            .address_to_validator_info
+            .iter()
+            .filter_map(|(address, consensus_info)| {
+                if authors.contains(address) {
+                    Some(&consensus_info.public_key)
+                } else {
+                    None
+                }
+            })
+            .collect();
+        let aggregated_key = PublicKey::aggregate(pub_keys_to_agg)
+            .map_err(|_| VerifyError::FailedToAggregatePubKey)?;
+
+        multi_signature
+            .multi_sig()
+            .as_ref()
+            .unwrap()
```

Comment:
need to revert the unwrap too, we should also followup with negative tests

Reply:
Have added a negative test case for signature verification - let me know if you have any other scenarios in mind. We can add them as a follow up.

Reply:
i was referring to the simple empty signature test